### DISCUSSION TITLE
How to send requests inside mitmproxy synchronously

### DISCUSSION THREAD

**Discussion by changsin** [mitmproxy/mitmproxy#6566] [General]
I tried to create an http request inside the mitmproxy but does not seem to work. Here is the gist of what I am doing:

import requests
...
    def request(self, flow: http.HTTPFlow) -> None:
        response = requests.post(url, headers={"X-Api-Key": API_KEY},
                                 json=event, timeout=timeout, proxies={})

The call times out. I see some related questions but they were trying to creating http.HTTPRequest object itself.

https://stackoverflow.com/questions/60957418/how-to-make-a-request-inside-a-simple-mitmproxy-script

I don't need to do that. I just need to make a simple http call.
If I make this async (fire-and-forget), then it works, but if I need to wait for the call to return then it never works.
My initial problem was not specifying an empty proxies parameter to prevent the call to go through mitmproxy again and thereby creating an infinite loop. But after adding the empty proxies, the call still doesn't return in sync mode.

**Comment by Prinzhorn** [comment]
How are you launching mitmproxy? Are you 100% sure that the request does not go through mitmproxy (transparent mode?)? Because it sounds exactly like that. It wouldn't produce an infinite loop, but your sync request would just lock up the entire proxy until it times out and the request hook finishes to give control back to the proxy core.

Can you provide a minimal repro? The exact command you use to start mitmproxy and a complete script. You can probably just make a request to example.com for demo purposes.

**Reply by changsin** [reply]
Here is the minimal repro script. You run by mitmproxy -s sync_test.py after the saving the script below as sync_test.py


```
import requests

from mitmproxy import http

TEST_URL = "https://example.com"


class MyMitmproxy:
    def request(self, flow: http.HTTPFlow) -> None:
        response = requests.get(TEST_URL, proxies={})
        print(response)

addons = [MyMitmproxy()]
```


If I call asynchronously like the following, the internet works but then I don't get the response back from GET TEST_URL either.

```
import requests

class MyMitmproxy:
    async def request(self, flow: http.HTTPFlow) -> None:
        loop = asyncio.get_event_loop()
        response = await loop.run_in_executor(None, requests.get, TEST_URL)
        print(f"{response.content}")


addons = [MyMitmproxy()]
```

**Reply by Prinzhorn** [reply]
Your sync script works as expected for me. I make a request via `curl --insecure --proxy localhost:8080 https://google.com`, I get the response from example.com and then curl gets the google response. I did use `mitmdump`  because where would you expect `print` to appear in the mitmproxy ui?

`<Response [200]>` here is the `print(response)`

```
$ mitmdump --scripts request.py 
[08:09:45.410] Loading script /tmp/mitm/request.py
[08:09:45.443] HTTP(S) proxy listening at *:8080.
[08:09:46.618][[::1]:40408] client connect
[08:09:46.656][[::1]:40408] server connect google.com:443 ([2a00:1450:4005:802::200e]:443)
<Response [200]>
[::1]:40408: GET https://google.com/ HTTP/2.0
 << HTTP/2.0 301 Moved Permanently 220b
[08:09:47.247][[::1]:40408] client disconnect
[08:09:47.248][[::1]:40408] server disconnect google.com:443 ([2a00:1450:4005:802::200e]:443)
```

Now if you also use `@concurrent` you won't block the proxy during the example request.

**Reply by changsin** [reply]
That is strange because when I execute the script, it does not return (the program hangs).
But I did find a solution for my problem. I used httpx instead to unblock myself. I found that requests block the current event loop while httpx does not.
Here is the gist of what I did and it worked.

```            with httpx.Client() as client:
                response = client.post(
                    url,
                    headers={"X-Api-Key": api_key},
                    json=event,
                    timeout=TIMEOUT_SECONDS
                )
```

**Reply by Prinzhorn** [reply]
With `httpx` you should be able to use `async def request` and then `await client.post` (with an `AsyncClient`). Then you're not blocking anything, as mhils said below.

**Comment by mhils** [comment]
You typically want to use an `async` HTTP library, otherwise you're blocking the entire proxy event loop. The only escape hatch is `@concurrent`, which comes with a bit of a performance overhead.

**Reply by changsin** [reply]
In our case, I want to wait till I get a response from a certain website and based on the response, we proceed or block the call. Is that why the following code doesn't work?

```
import requests

from mitmproxy import http

TEST_URL = "https://example.com/"

class MyMitmproxy:
    def request(self, flow: http.HTTPFlow) -> None:
        response = requests.get(TEST_URL, proxies={})
        print(response)

addons = [MyMitmproxy()]
```

I also tried using concurrent library you mentioned. Is this what you mean?

```
import asyncio
import concurrent.futures
import logging

import requests

from mitmproxy import http
from mitmproxy import options
from mitmproxy.tools import dump


TEST_URL = "https://example.com"
LOCAL_HOST = "127.0.0.1"
DEFAULT_PORT = 8080


class MyMitmproxy:
    def get_result(self, url):
        response = requests.get(url)
        return response

    def request(self, flow: http.HTTPFlow) -> None:
        # Using ThreadPoolExecutor for parallel execution
        with concurrent.futures.ThreadPoolExecutor() as executor:
            future1 = executor.submit(self.get_result, TEST_URL)

            # Retrieve results when ready
            result1 = future1.result()
            print(f"###result is {result1}")


async def run_mitmproxy_async(mymitmproxy : MyMitmproxy, opts):
    master_instance = dump.DumpMaster(
        opts,
        with_termlog=False,
        with_dumper=False,
    )
    master_instance.addons.add(mymitmproxy)

    await master_instance.run()


async def run_mitmproxy_task(mymitmproxy : MyMitmproxy, host=LOCAL_HOST, port=DEFAULT_PORT):
    opts = options.Options(listen_host=host, listen_port=port,
                           confdir="/etc/ssl/certs")
    mitm_task = asyncio.create_task(run_mitmproxy_async(mymitmproxy, opts))
    await mitm_task


if __name__ == '__main__':
    asyncio.run(run_mitmproxy_task(MyMitmproxy()))
```

**Comment by Malesenka** [comment]
thanks for the repository
